### PR TITLE
[Merged by Bors] - feat(Algebra/GroupWithZero): MulEquiv between WithZero (units) and the base group with zero

### PR DESCRIPTION
**Pull request by pechersky**
Also MulEquivs that add or remove WithZero on both sides 
And some basic WithOne/WithZero API

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/GroupWithZero/WithZero.lean [hunk 1]
```diff
@@ -266,6 +266,45 @@ def unitsWithZeroEquiv : (WithZero α)ˣ ≃* α where
...
+  toFun := map' e.toMonoidHom
+  invFun := map' e.symm.toMonoidHom
+  left_inv := (by induction · <;> simp)
+  right_inv := (by induction · <;> simp)
+  map_mul' := by
+    intro x y
+    induction x <;> induction y <;>
+    simp
+
+/-- The inverse of `MulEquiv.withZeroCongr`. -/
```
Comment (YaelDillies) [inline Mathlib/Algebra/GroupWithZero/WithZero.lean:293]:
```suggestion
/-- The inverse of `MulEquiv.withZero`. -/
```

File: Mathlib/Algebra/GroupWithZero/WithZero.lean [hunk 2]
```diff
@@ -266,6 +266,45 @@ def unitsWithZeroEquiv : (WithZero α)ˣ ≃* α where
...
+protected noncomputable def _root_.MulEquiv.unzero [Group β] (e : WithZero α ≃* WithZero β) :
+    α ≃* β where
+  toFun x := unzero (x := e x) (by simp [ne_eq, ← e.eq_symm_apply])
+  invFun x := unzero (x := e.symm x) (by simp [e.symm_apply_eq])
+  left_inv _ := by simp
+  right_inv _ := by simp
+  map_mul' := by
+    intro x y
```
Comment (YaelDillies) [inline Mathlib/Algebra/GroupWithZero/WithZero.lean:301]:
```suggestion
  map_mul' x y := by
```

File: Mathlib/Algebra/GroupWithZero/WithZero.lean [hunk 3]
```diff
@@ -266,6 +266,45 @@ def unitsWithZeroEquiv : (WithZero α)ˣ ≃* α where
...
+    simp only [coe_mul, map_mul]
+    generalize_proofs A B C
+    suffices (((unzero A) : β) : WithZero β) = (unzero B) * (unzero C) by
```
Comment (YaelDillies) [inline Mathlib/Algebra/GroupWithZero/WithZero.lean:304]:
```suggestion
    suffices ((unzero A : β) : WithZero β) = unzero B * unzero C by
```

File: Mathlib/Algebra/GroupWithZero/WithZero.lean [hunk 4]
```diff
@@ -266,6 +266,44 @@ def unitsWithZeroEquiv : (WithZero α)ˣ ≃* α where
...
+
+/-- A version of `Equiv.optionCongr` for `WithZero`. -/
+noncomputable def _root_.MulEquiv.withZero [Group β] (e : α ≃* β) :
+    WithZero α ≃* WithZero β where
+  toFun := map' e.toMonoidHom
+  invFun := map' e.symm.toMonoidHom
+  left_inv := (by induction · <;> simp)
+  right_inv := (by induction · <;> simp)
+  map_mul' := by
+    intro x y
```
Comment (YaelDillies) [inline Mathlib/Algebra/GroupWithZero/WithZero.lean:289]:
```suggestion
  map_mul' x y := by
```

**Review by YaelDillies — COMMENTED** [review]
Apart from that,

maintainer merge

**Comment by fpvandoorn** [comment]
bors merge

**Pull request closed without merging**